socket.c

```diff
@@ -765,13 +768,96 @@ static opa_socket_context socket_eval(const char *input)
 	return this_cpu_opa_socket_eval(input);
 }
 
+static command_answer *prepare_opa_input(direction direction, struct sock *s, u16 port, char *attest_response_json)
+{
+	const char *ipformat = "%pI4";
+
+	if (s->sk_family == AF_INET6)
+	{
+		ipformat = "%pI6";
+	}
+
+	char source_ip[INET6_ADDRSTRLEN];
+	u16 source_port;
+	char destination_ip[INET6_ADDRSTRLEN];
+	u16 destination_port;
+
+	if (direction == INPUT)
+	{
+		snprintf(source_ip, INET6_ADDRSTRLEN, ipformat, &s->sk_daddr);
+		snprintf(destination_ip, INET6_ADDRSTRLEN, ipformat, &s->sk_rcv_saddr);
+		source_port = s->sk_dport;
+		destination_port = s->sk_num;
+	}
+	else
+	{
+		snprintf(source_ip, INET6_ADDRSTRLEN, ipformat, &s->sk_rcv_saddr);
+		snprintf(destination_ip, INET6_ADDRSTRLEN, ipformat, &s->sk_daddr);
+		source_port = s->sk_num;
+		destination_port = port;
+	}
+
+	if (!attest_response_json)
+	{
+		return answer_with_error("nil attest response json");
+	}
+
+	JSON_Value *json = json_parse_string(attest_response_json);
+	if (!json)
+	{
+		return answer_with_error("could not parse json");
+	}
+
+	JSON_Object *root = json_value_get_object(json);
+	if (!root)
+	{
+		return answer_with_error("could not get root object");
+	}
+
+	char buff[256];
+
+	JSON_Object *selectors = json_object_get_object(root, "selectors");
+	if (!selectors)
+	{
+		return answer_with_error("could not find selectors in json");
+	}
+
+	if (direction == INPUT)
+		json_object_set_boolean(selectors, "direction:input", true);
+	else
+		json_object_set_boolean(selectors, "direction:output", true);
+
+	snprintf(buff, sizeof(buff), "source:ip:%s", source_ip);
+	json_object_set_boolean(selectors, buff, true);
+	snprintf(buff, sizeof(buff), "source:port:%d", source_port);
+	json_object_set_boolean(selectors, buff, true);
+
+	snprintf(buff, sizeof(buff), "destination:ip:%s", destination_ip);
+	json_object_set_boolean(selectors, buff, true);
+	snprintf(buff, sizeof(buff), "destination:port:%d", destination_port);
+	json_object_set_boolean(selectors, buff, true);
+
+	char *serialized_string = json_serialize_to_string(json);
+
+	command_answer *answer = kzalloc(sizeof(struct command_answer), GFP_KERNEL);
+	answer->answer = strdup(serialized_string);
```

Comment:
No need to dup the string `serialized_string` here, just pass it as the answer, and we can remove the `json_free_serialized_string` call from here as well (which is just a `kfree`).